src/Compression/Compression.cs

```diff
@@ -300,11 +300,37 @@ private static void ThrowCannotWriteToDeflateStreamException()
 			throw new InvalidOperationException ("Writing to the compression stream is not supported.");
 		}
 
-		public override IAsyncResult BeginRead (byte[] buffer, int offset, int count, AsyncCallback asyncCallback, object asyncState) =>
-			TaskToApm.Begin(ReadAsync (buffer, offset, count, CancellationToken.None), asyncCallback, asyncState);
-
-		public override int EndRead (IAsyncResult asyncResult) =>
-			TaskToApm.End<int> (asyncResult);
+#if DOTNET
+		#pragma warning disable 0809 // Obsolete member 'CompressionStream.BeginRead(...)' overrides non-obsolete member 'Stream.BeginRead(...)'
+		[Obsolete ("Use 'ReadAsync' instead. This method is synchronous.")]
+#endif
+		public override IAsyncResult BeginRead (byte[] buffer, int offset, int count, AsyncCallback asyncCallback, object asyncState)
+		{
```

Comment:
A similar Mono change recently [broke our product](https://github.com/mono/mono/issues/18865). This is quite problematic because the base implementation of `BeginRead`/`BeginWrite`/`EndRead`/`EndWrite`. It is calling `ReadAsync`/`WriteAsync` internally but it is geared towards seekable streams where `Position` matters and it only allows single async call at a time. This is particularly problematic with [wrapped] network streams where a common coding pattern expect both async read and write to be working at the same time.

Note that in .NET Core all non-seekable network streams or streams that wrap around network streams (GzipStream, HttpBaseStream, QuicStream, PipeReaderStream, PipeWriterStream, ReadOnlyMemoryStream, BrotliStream, SslStream, ...) use the `TaskToApm` pattern.

I think the base implementation in `Stream.BeginRead/BeginWrite` could probably be fixed/extended to work for these scenarios but that's not the case today. Until that happens I think it's better to use a private copy of `TaskToApm` here.

Reply:
I filed an issue at https://github.com/dotnet/runtime/issues/33447 to seek better solution. If it gains some traction it may make sense to remove the overrides of BeginRead/BeginWrite/EndRead/EndWrite altogether.